// terra-api/dist/cjs/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.integrations = exports.plannedworkout = exports.sleep = exports.nutrition = exports.menstruation = exports.daily = exports.body = exports.athlete = exports.activity = exports.user = exports.authentication = void 0;
exports.authentication = __importStar(require("./authentication/index.js"));
__exportStar(require("./authentication/types/index.js"), exports);
exports.user = __importStar(require("./user/index.js"));
__exportStar(require("./user/types/index.js"), exports);
exports.activity = __importStar(require("./activity/index.js"));
__exportStar(require("./activity/types/index.js"), exports);
exports.athlete = __importStar(require("./athlete/index.js"));
__exportStar(require("./athlete/types/index.js"), exports);
exports.body = __importStar(require("./body/index.js"));
__exportStar(require("./body/types/index.js"), exports);
exports.daily = __importStar(require("./daily/index.js"));
__exportStar(require("./daily/types/index.js"), exports);
exports.menstruation = __importStar(require("./menstruation/index.js"));
__exportStar(require("./menstruation/types/index.js"), exports);
exports.nutrition = __importStar(require("./nutrition/index.js"));
__exportStar(require("./nutrition/types/index.js"), exports);
exports.sleep = __importStar(require("./sleep/index.js"));
__exportStar(require("./sleep/types/index.js"), exports);
exports.plannedworkout = __importStar(require("./plannedworkout/index.js"));
__exportStar(require("./plannedworkout/types/index.js"), exports);
exports.integrations = __importStar(require("./integrations/index.js"));
__exportStar(require("./integrations/types/index.js"), exports);
__exportStar(require("./authentication/client/requests/index.js"), exports);
__exportStar(require("./user/client/requests/index.js"), exports);
__exportStar(require("./activity/client/requests/index.js"), exports);
__exportStar(require("./athlete/client/requests/index.js"), exports);
__exportStar(require("./body/client/requests/index.js"), exports);
__exportStar(require("./daily/client/requests/index.js"), exports);
__exportStar(require("./menstruation/client/requests/index.js"), exports);
__exportStar(require("./nutrition/client/requests/index.js"), exports);
__exportStar(require("./sleep/client/requests/index.js"), exports);
__exportStar(require("./plannedworkout/client/requests/index.js"), exports);
__exportStar(require("./integrations/client/requests/index.js"), exports);
