// terra-api/dist/esm/api/types/CardioPlannedWorkoutStep.mjs
/**
 * This file was auto-generated by Fern from our API Definition.
 */
export var CardioPlannedWorkoutStep;
(function (CardioPlannedWorkoutStep) {
    CardioPlannedWorkoutStep.ExerciseCategory = {
        Unknown: "UNKNOWN",
        BenchPress: "BENCH_PRESS",
        CalfRaise: "CALF_RAISE",
        Cardio: "CARDIO",
        Carry: "CARRY",
        Chop: "CHOP",
        Core: "CORE",
        Crunch: "CRUNCH",
        Curl: "CURL",
        Deadlift: "DEADLIFT",
        Flye: "FLYE",
        HipRaise: "HIP_RAISE",
        HipStability: "HIP_STABILITY",
        HipSwing: "HIP_SWING",
        Hyperextension: "HYPEREXTENSION",
        LateralRaise: "LATERAL_RAISE",
        LegCurl: "LEG_CURL",
        LegRaise: "LEG_RAISE",
        Lunge: "LUNGE",
        OlympicLift: "OLYMPIC_LIFT",
        Plank: "PLANK",
        Plyo: "PLYO",
        PullUp: "PULL_UP",
        PushUp: "PUSH_UP",
        Row: "ROW",
        ShoulderPress: "SHOULDER_PRESS",
        ShoulderStability: "SHOULDER_STABILITY",
        Shrug: "SHRUG",
        SitUp: "SIT_UP",
        Squat: "SQUAT",
        TotalBody: "TOTAL_BODY",
        TricepsExtension: "TRICEPS_EXTENSION",
        WarmUp: "WARM_UP",
        Run: "RUN",
        Bike: "BIKE",
        CardioSensors: "CARDIO_SENSORS",
    };
})(CardioPlannedWorkoutStep || (CardioPlannedWorkoutStep = {}));
