// terra-api/dist/cjs/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraClient = void 0;
const core = __importStar(require("./core/index.js"));
const headers_js_1 = require("./core/headers.js");
const Client_js_1 = require("./api/resources/authentication/client/Client.js");
const Client_js_2 = require("./api/resources/user/client/Client.js");
const Client_js_3 = require("./api/resources/activity/client/Client.js");
const Client_js_4 = require("./api/resources/athlete/client/Client.js");
const Client_js_5 = require("./api/resources/body/client/Client.js");
const Client_js_6 = require("./api/resources/daily/client/Client.js");
const Client_js_7 = require("./api/resources/menstruation/client/Client.js");
const Client_js_8 = require("./api/resources/nutrition/client/Client.js");
const Client_js_9 = require("./api/resources/sleep/client/Client.js");
const Client_js_10 = require("./api/resources/plannedworkout/client/Client.js");
const Client_js_11 = require("./api/resources/integrations/client/Client.js");
class TerraClient {
    constructor(_options) {
        this._options = Object.assign(Object.assign({}, _options), { headers: (0, headers_js_1.mergeHeaders)({
                "dev-id": _options === null || _options === void 0 ? void 0 : _options.devId,
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "terra-api",
                "X-Fern-SDK-Version": "2.1.0",
                "User-Agent": "terra-api/2.1.0",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get authentication() {
        var _a;
        return ((_a = this._authentication) !== null && _a !== void 0 ? _a : (this._authentication = new Client_js_1.Authentication(this._options)));
    }
    get user() {
        var _a;
        return ((_a = this._user) !== null && _a !== void 0 ? _a : (this._user = new Client_js_2.User(this._options)));
    }
    get activity() {
        var _a;
        return ((_a = this._activity) !== null && _a !== void 0 ? _a : (this._activity = new Client_js_3.Activity(this._options)));
    }
    get athlete() {
        var _a;
        return ((_a = this._athlete) !== null && _a !== void 0 ? _a : (this._athlete = new Client_js_4.Athlete(this._options)));
    }
    get body() {
        var _a;
        return ((_a = this._body) !== null && _a !== void 0 ? _a : (this._body = new Client_js_5.Body(this._options)));
    }
    get daily() {
        var _a;
        return ((_a = this._daily) !== null && _a !== void 0 ? _a : (this._daily = new Client_js_6.Daily(this._options)));
    }
    get menstruation() {
        var _a;
        return ((_a = this._menstruation) !== null && _a !== void 0 ? _a : (this._menstruation = new Client_js_7.Menstruation(this._options)));
    }
    get nutrition() {
        var _a;
        return ((_a = this._nutrition) !== null && _a !== void 0 ? _a : (this._nutrition = new Client_js_8.Nutrition(this._options)));
    }
    get sleep() {
        var _a;
        return ((_a = this._sleep) !== null && _a !== void 0 ? _a : (this._sleep = new Client_js_9.Sleep(this._options)));
    }
    get plannedworkout() {
        var _a;
        return ((_a = this._plannedworkout) !== null && _a !== void 0 ? _a : (this._plannedworkout = new Client_js_10.Plannedworkout(this._options)));
    }
    get integrations() {
        var _a;
        return ((_a = this._integrations) !== null && _a !== void 0 ? _a : (this._integrations = new Client_js_11.Integrations(this._options)));
    }
}
exports.TerraClient = TerraClient;
