// terra-api/dist/cjs/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookVerificationError = void 0;
exports.verifyTerraWebhookSignature = verifyTerraWebhookSignature;
/**
 * Custom error class for webhook verification failures
 */
class WebhookVerificationError extends Error {
    constructor(message) {
        super(message);
        this.name = "WebhookVerificationError";
    }
}
exports.WebhookVerificationError = WebhookVerificationError;
/**
 * Verify Terra webhook signature using HMAC-SHA256.
 *
 * @param payload - The raw JSON payload from the request body
 * @param signatureHeader - The 'terra-signature' header value
 * @param signingSecret - The webhook endpoint's signing secret
 * @param options - Configuration options including tolerance
 * @returns A Promise resolving to true if the signature is valid
 * @throws WebhookVerificationError if verification fails
 */
function verifyTerraWebhookSignature(payload_1, signatureHeader_1, signingSecret_1) {
    return __awaiter(this, arguments, void 0, function* (payload, signatureHeader, signingSecret, options = {}) {
        let crypto;
        try {
            crypto = yield Promise.resolve().then(() => __importStar(require("crypto")));
        }
        catch (error) {
            throw new WebhookVerificationError("Webhook verification requires the Node.js 'crypto' module, which could not be loaded.");
        }
        const { tolerance = 300 } = options;
        if (!payload || !signatureHeader || !signingSecret) {
            throw new WebhookVerificationError("Missing required parameters");
        }
        const { timestamp, signatures } = extractTimestampAndSignatures(signatureHeader);
        if (!timestamp) {
            throw new WebhookVerificationError("No timestamp found in signature header");
        }
        if (signatures.length === 0) {
            throw new WebhookVerificationError("No v1 signatures found in signature header");
        }
        const signedPayload = `${timestamp}.${payload}`;
        // Pass the crypto module to the helper function
        const expectedSignature = computeSignature(signedPayload, signingSecret, crypto);
        // Pass the timingSafeEqual function to the helper function
        verifySignatures(signatures, expectedSignature, crypto.timingSafeEqual);
        verifyTimestamp(timestamp, tolerance);
        return true;
    });
}
/**
 * Extract timestamp and v1 signatures from the terra-signature header.
 */
function extractTimestampAndSignatures(signatureHeader) {
    let timestamp = null;
    const signatures = [];
    for (const element of signatureHeader.split(",")) {
        const trimmedElement = element.trim();
        if (!trimmedElement.includes("=")) {
            continue;
        }
        const [prefix, ...valueParts] = trimmedElement.split("=");
        const value = valueParts.join("=").trim();
        if (prefix.trim() === "t") {
            timestamp = value;
        }
        else if (prefix.trim() === "v1") {
            signatures.push(value);
        }
    }
    return { timestamp, signatures };
}
/**
 * Compute HMAC-SHA256 signature for the signed payload.
 * @param signedPayload - The signed payload
 * @param signingSecret - The webhook endpoint's signing secret
 * @param crypto - Node.js crypto module
 */
function computeSignature(signedPayload, signingSecret, crypto) {
    return crypto.createHmac("sha256", signingSecret).update(signedPayload, "utf8").digest("hex");
}
/**
 * Verify signatures using a timing-safe comparison.
 * @param receivedSignatures - The received signatures
 * @param expectedSignature - The expected signature
 * @param timingSafeEqual - The crypto.timingSafeEqual function
 */
function verifySignatures(receivedSignatures, expectedSignature, timingSafeEqual) {
    const expectedBuffer = Buffer.from(expectedSignature, "hex");
    for (const signature of receivedSignatures) {
        try {
            const signatureBuffer = Buffer.from(signature, "hex");
            if (expectedBuffer.length === signatureBuffer.length && timingSafeEqual(expectedBuffer, signatureBuffer)) {
                return; // Signature is valid
            }
        }
        catch (error) {
            // Invalid hex string
            console.error("Invalid hex string", error);
        }
    }
    throw new WebhookVerificationError("No matching signature found");
}
/**
 * Verify that the timestamp is within the acceptable tolerance.
 */
function verifyTimestamp(timestampStr, tolerance) {
    const webhookTimestamp = parseInt(timestampStr, 10);
    if (isNaN(webhookTimestamp)) {
        throw new WebhookVerificationError("Invalid timestamp format");
    }
    const currentTimestamp = Math.floor(Date.now() / 1000);
    const age = currentTimestamp - webhookTimestamp;
    if (age > tolerance) {
        throw new WebhookVerificationError(`Webhook timestamp too old: ${age}s > ${tolerance}s tolerance`);
    }
    if (age < -tolerance) {
        throw new WebhookVerificationError(`Webhook timestamp too far in future: ${Math.abs(age)}s > ${tolerance}s tolerance`);
    }
}
