// terra-api/dist/cjs/api/types/PlannedWorkoutMetadata.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlannedWorkoutMetadata = void 0;
var PlannedWorkoutMetadata;
(function (PlannedWorkoutMetadata) {
    PlannedWorkoutMetadata.Type = {
        InVehicle: "IN_VEHICLE",
        Biking: "BIKING",
        Still: "STILL",
        Unknown: "UNKNOWN",
        Tilting: "TILTING",
        Walking: "WALKING",
        Running: "RUNNING",
        Aerobics: "AEROBICS",
        Badminton: "BADMINTON",
        Baseball: "BASEBALL",
        Basketball: "BASKETBALL",
        Biathlon: "BIATHLON",
        Handbiking: "HANDBIKING",
        MountainBiking: "MOUNTAIN_BIKING",
        RoadBiking: "ROAD_BIKING",
        Spinning: "SPINNING",
        StationaryBiking: "STATIONARY_BIKING",
        UtilityBiking: "UTILITY_BIKING",
        Boxing: "BOXING",
        Calisthenics: "CALISTHENICS",
        CircuitTraining: "CIRCUIT_TRAINING",
        Cricket: "CRICKET",
        Dancing: "DANCING",
        Elliptical: "ELLIPTICAL",
        Fencing: "FENCING",
        AmericanFootball: "AMERICAN_FOOTBALL",
        AustralianFootball: "AUSTRALIAN_FOOTBALL",
        EnglishFootball: "ENGLISH_FOOTBALL",
        Frisbee: "FRISBEE",
        Gardening: "GARDENING",
        Golf: "GOLF",
        Gymnastics: "GYMNASTICS",
        Handball: "HANDBALL",
        Hiking: "HIKING",
        Hockey: "HOCKEY",
        HorsebackRiding: "HORSEBACK_RIDING",
        Housework: "HOUSEWORK",
        JumpingRope: "JUMPING_ROPE",
        Kayaking: "KAYAKING",
        KettlebellTraining: "KETTLEBELL_TRAINING",
        Kickboxing: "KICKBOXING",
        Kitesurfing: "KITESURFING",
        MartialArts: "MARTIAL_ARTS",
        Meditation: "MEDITATION",
        MixedMartialArts: "MIXED_MARTIAL_ARTS",
        P90XExercises: "P90X_EXERCISES",
        Paragliding: "PARAGLIDING",
        Pilates: "PILATES",
        Polo: "POLO",
        Racquetball: "RACQUETBALL",
        RockClimbing: "ROCK_CLIMBING",
        Rowing: "ROWING",
        RowingMachine: "ROWING_MACHINE",
        Rugby: "RUGBY",
        Jogging: "JOGGING",
        RunningOnSand: "RUNNING_ON_SAND",
        TreadmillRunning: "TREADMILL_RUNNING",
        Sailing: "SAILING",
        ScubaDiving: "SCUBA_DIVING",
        Skateboarding: "SKATEBOARDING",
        Skating: "SKATING",
        CrossSkating: "CROSS_SKATING",
        IndoorRollerblading: "INDOOR_ROLLERBLADING",
        Skiing: "SKIING",
        BackCountrySkiing: "BACK_COUNTRY_SKIING",
        CrossCountrySkiing: "CROSS_COUNTRY_SKIING",
        DownhillSkiing: "DOWNHILL_SKIING",
        KiteSkiing: "KITE_SKIING",
        RollerSkiing: "ROLLER_SKIING",
        Sledding: "SLEDDING",
        Snowboarding: "SNOWBOARDING",
        Snowmobile: "SNOWMOBILE",
        Snowshoeing: "SNOWSHOEING",
        Squash: "SQUASH",
        StairClimbing: "STAIR_CLIMBING",
        StairClimbingMachine: "STAIR_CLIMBING_MACHINE",
        StandUpPaddleboarding: "STAND_UP_PADDLEBOARDING",
        StrengthTraining: "STRENGTH_TRAINING",
        Surfing: "SURFING",
        Swimming: "SWIMMING",
        SwimmingSwimmingPool: "SWIMMING_SWIMMING_POOL",
        SwimmingOpenWater: "SWIMMING_OPEN_WATER",
        TableTennis: "TABLE_TENNIS",
        TeamSports: "TEAM_SPORTS",
        Tennis: "TENNIS",
        Treadmill: "TREADMILL",
        Volleyball: "VOLLEYBALL",
        VolleyballBeach: "VOLLEYBALL_BEACH",
        VolleyballIndoor: "VOLLEYBALL_INDOOR",
        Wakeboarding: "WAKEBOARDING",
        WalkingFitness: "WALKING_FITNESS",
        NordicWalking: "NORDIC_WALKING",
        WalkingTreadmill: "WALKING_TREADMILL",
        Waterpolo: "WATERPOLO",
        Weightlifting: "WEIGHTLIFTING",
        Wheelchair: "WHEELCHAIR",
        Windsurfing: "WINDSURFING",
        Yoga: "YOGA",
        Zumba: "ZUMBA",
        Diving: "DIVING",
        Ergometer: "ERGOMETER",
        IceSkating: "ICE_SKATING",
        IndoorSkating: "INDOOR_SKATING",
        Curling: "CURLING",
        Other: "OTHER",
        Crossfit: "CROSSFIT",
        Hiit: "HIIT",
        IntervalTraining: "INTERVAL_TRAINING",
        WalkingStroller: "WALKING_STROLLER",
        Elevator: "ELEVATOR",
        Escalator: "ESCALATOR",
        Archery: "ARCHERY",
        Softball: "SOFTBALL",
        GuidedBreathing: "GUIDED_BREATHING",
        CardioTraining: "CARDIO_TRAINING",
        Lacrosse: "LACROSSE",
        Stretching: "STRETCHING",
        Triathlon: "TRIATHLON",
        InlineSkating: "INLINE_SKATING",
        SkyDiving: "SKY_DIVING",
        Paddling: "PADDLING",
        Mountaineering: "MOUNTAINEERING",
        Fishing: "FISHING",
        WaterSkiing: "WATER_SKIING",
        IndoorRunning: "INDOOR_RUNNING",
    };
})(PlannedWorkoutMetadata || (exports.PlannedWorkoutMetadata = PlannedWorkoutMetadata = {}));
