// terra-api/dist/esm/api/resources/activity/client/Client.mjs
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as environments from "../../../../environments.mjs";
import * as core from "../../../../core/index.mjs";
import * as Terra from "../../../index.mjs";
import { toJson } from "../../../../core/json.mjs";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.mjs";
import * as errors from "../../../../errors/index.mjs";
export class Activity {
    constructor(_options) {
        this._options = _options;
    }
    /**
     * Fetches completed workout sessions, with a defined start and end time and activity type (e.g. running, cycling, etc.)
     *
     * @param {Terra.ActivityFetchRequest} request
     * @param {Activity.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Terra.BadRequestError}
     * @throws {@link Terra.UnauthorizedError}
     * @throws {@link Terra.NotFoundError}
     *
     * @example
     *     await client.activity.fetch({
     *         user_id: "user_id",
     *         start_date: 1,
     *         end_date: 1,
     *         to_webhook: true,
     *         with_samples: true
     *     })
     */
    fetch(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__fetch(request, requestOptions));
    }
    __fetch(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e;
            const { user_id: userId, start_date: startDate, end_date: endDate, to_webhook: toWebhook, with_samples: withSamples, } = request;
            const _queryParams = {};
            _queryParams["user_id"] = userId;
            _queryParams["start_date"] = typeof startDate === "string" ? startDate : toJson(startDate);
            if (endDate != null) {
                _queryParams["end_date"] = typeof endDate === "string" ? endDate : toJson(endDate);
            }
            if (toWebhook != null) {
                _queryParams["to_webhook"] = toWebhook.toString();
            }
            if (withSamples != null) {
                _queryParams["with_samples"] = withSamples.toString();
            }
            let _headers = mergeHeaders((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, mergeOnlyDefinedHeaders(Object.assign({ "dev-id": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.devId) !== null && _b !== void 0 ? _b : (_c = this._options) === null || _c === void 0 ? void 0 : _c.devId }, (yield this._getCustomAuthorizationHeaders()))), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.TerraEnvironment.Default, "activity"),
                method: "GET",
                headers: _headers,
                queryParameters: Object.assign(Object.assign({}, _queryParams), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams),
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Terra.BadRequestError(_response.error.body, _response.rawResponse);
                    case 401:
                        throw new Terra.UnauthorizedError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Terra.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TerraError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TerraError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TerraTimeoutError("Timeout exceeded when calling GET /activity.");
                case "unknown":
                    throw new errors.TerraError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    /**
     * Used to post activity data to a provider. This endpoint only works for users connected via Wahoo. Returns error for other providers.
     *
     * @param {Terra.ActivityWriteRequest} request
     * @param {Activity.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Terra.BadRequestError}
     * @throws {@link Terra.NotFoundError}
     *
     * @example
     *     await client.activity.write({
     *         data: [{
     *                 metadata: {
     *                     end_time: "2022-10-28T10:00:00.000000+01:00",
     *                     start_time: "1999-11-23T09:00:00.000000+02:00",
     *                     summary_id: "123e4567-e89b-12d3-a456-426614174000",
     *                     type: 1.1,
     *                     upload_type: 1.1
     *                 }
     *             }]
     *     })
     */
    write(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__write(request, requestOptions));
    }
    __write(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e;
            let _headers = mergeHeaders((_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, mergeOnlyDefinedHeaders(Object.assign({ "dev-id": (_b = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.devId) !== null && _b !== void 0 ? _b : (_c = this._options) === null || _c === void 0 ? void 0 : _c.devId }, (yield this._getCustomAuthorizationHeaders()))), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_e = (_d = (yield core.Supplier.get(this._options.baseUrl))) !== null && _d !== void 0 ? _d : (yield core.Supplier.get(this._options.environment))) !== null && _e !== void 0 ? _e : environments.TerraEnvironment.Default, "activity"),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryParameters: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams,
                requestType: "json",
                body: request,
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Terra.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Terra.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.TerraError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.TerraError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.TerraTimeoutError("Timeout exceeded when calling POST /activity.");
                case "unknown":
                    throw new errors.TerraError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    _getCustomAuthorizationHeaders() {
        return __awaiter(this, void 0, void 0, function* () {
            const apiKeyValue = yield core.Supplier.get(this._options.apiKey);
            return { "x-api-key": apiKeyValue };
        });
    }
}
