// terra-api/dist/esm/Client.mjs
/**
 * This file was auto-generated by Fern from our API Definition.
 */
import * as core from "./core/index.mjs";
import { mergeHeaders } from "./core/headers.mjs";
import { Authentication } from "./api/resources/authentication/client/Client.mjs";
import { User } from "./api/resources/user/client/Client.mjs";
import { Activity } from "./api/resources/activity/client/Client.mjs";
import { Athlete } from "./api/resources/athlete/client/Client.mjs";
import { Body } from "./api/resources/body/client/Client.mjs";
import { Daily } from "./api/resources/daily/client/Client.mjs";
import { Menstruation } from "./api/resources/menstruation/client/Client.mjs";
import { Nutrition } from "./api/resources/nutrition/client/Client.mjs";
import { Sleep } from "./api/resources/sleep/client/Client.mjs";
import { Plannedworkout } from "./api/resources/plannedworkout/client/Client.mjs";
import { Integrations } from "./api/resources/integrations/client/Client.mjs";
export class TerraClient {
    constructor(_options) {
        this._options = Object.assign(Object.assign({}, _options), { headers: mergeHeaders({
                "dev-id": _options === null || _options === void 0 ? void 0 : _options.devId,
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "terra-api",
                "X-Fern-SDK-Version": "2.1.0",
                "User-Agent": "terra-api/2.1.0",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get authentication() {
        var _a;
        return ((_a = this._authentication) !== null && _a !== void 0 ? _a : (this._authentication = new Authentication(this._options)));
    }
    get user() {
        var _a;
        return ((_a = this._user) !== null && _a !== void 0 ? _a : (this._user = new User(this._options)));
    }
    get activity() {
        var _a;
        return ((_a = this._activity) !== null && _a !== void 0 ? _a : (this._activity = new Activity(this._options)));
    }
    get athlete() {
        var _a;
        return ((_a = this._athlete) !== null && _a !== void 0 ? _a : (this._athlete = new Athlete(this._options)));
    }
    get body() {
        var _a;
        return ((_a = this._body) !== null && _a !== void 0 ? _a : (this._body = new Body(this._options)));
    }
    get daily() {
        var _a;
        return ((_a = this._daily) !== null && _a !== void 0 ? _a : (this._daily = new Daily(this._options)));
    }
    get menstruation() {
        var _a;
        return ((_a = this._menstruation) !== null && _a !== void 0 ? _a : (this._menstruation = new Menstruation(this._options)));
    }
    get nutrition() {
        var _a;
        return ((_a = this._nutrition) !== null && _a !== void 0 ? _a : (this._nutrition = new Nutrition(this._options)));
    }
    get sleep() {
        var _a;
        return ((_a = this._sleep) !== null && _a !== void 0 ? _a : (this._sleep = new Sleep(this._options)));
    }
    get plannedworkout() {
        var _a;
        return ((_a = this._plannedworkout) !== null && _a !== void 0 ? _a : (this._plannedworkout = new Plannedworkout(this._options)));
    }
    get integrations() {
        var _a;
        return ((_a = this._integrations) !== null && _a !== void 0 ? _a : (this._integrations = new Integrations(this._options)));
    }
}
